src/gateway/routers/Aggregator.sol

```diff
@@ -91,42 +91,37 @@ contract Aggregator is Auth, IAggregator {
         }
 
         bytes32 messageHash;
-        ConfirmationState storage state;
         if (isMessageProof) {
+            require(isRecovery || router.id != 1, "RouterAggregator/non-proof-router");
             messageHash = MessagesLib.parseMessageProof(payload);
-            state = _confirmations[messageHash];
-            state.proofs[router.id - 1]++;
-
             emit HandleProof(messageHash, msg.sender);
         } else {
+            require(isRecovery || router.id == 1, "RouterAggregator/non-message-router");
             messageHash = keccak256(payload);
-            state = _confirmations[messageHash];
-            state.messages[router.id - 1]++;
-
             emit HandleMessage(payload, msg.sender);
         }
 
-        if (state.messages.countNonZeroValues() >= 1 && state.proofs.countNonZeroValues() >= router.quorum - 1) {
-            // Reduce total message confirmation count by 1, by finding the first non-zero value
-            state.messages.decreaseFirstNValues(1, 1);
+        Message storage state = messages[messageHash];
+        state.votes[router.id - 1]++;
 
-            // Reduce total proof confiration count by quorum
-            state.proofs.decreaseFirstNValues(router.quorum, 1);
+        if (state.votes.countNonZeroValues() >= router.quorum) {
+            // Reduce votes by quorum
+            state.votes.decreaseFirstNValues(router.quorum);
 
             if (isMessageProof) {
-                gateway.handle(pendingMessages[messageHash]);
-
-                // Only if there are no more pending messages, remove the pending message
-                if (state.messages.isEmpty() && state.proofs.isEmpty()) {
-                    delete pendingMessages[messageHash];
-                }
+                gateway.handle(state.pendingMessage);
             } else {
                 gateway.handle(payload);
             }
 
+            // Only if there are no more pending messages, remove the pending message
+            if (state.votes.isEmpty()) {
+                delete state.pendingMessage;
```

Comment:
I don't think there's any risk to just leaving the pendingMessage is there? We could save a little gas by skipping this.

Reply:
You get gas refunds if you clear storage, so this actually is a gas optimization

Reply:
Oh wow! Didn't know that